### PR TITLE
Readme for docker, dependency update, JDK13 support

### PR DESCRIPTION
**Pull request by mfbieber**
**Adapted dependencies** in pom.xml (newer versions of cqengine, kubernetes-client and kubernetes-server-mock) to reduce warnings in build:

- spring-boot-starter-parent, version 2.2.7 RELEASE has now a different location in the artifactory (release instead of lib-release): https://repo.spring.io/webapp/#/artifacts/browse/tree/General/release/org/springframework/boot/spring-boot-starter-parent
- kubernetes-client and kubernetes-server-mock have dependencies to sundr-codegen which causes the build to display these warnings:
 ```
The POM for io.sundr:sundr-codegen:jar:0.13.1 is invalid, transitive dependencies (if any) will not be available: 2 problems were encountered while building the effective model for io.sundr:sundr-codegen:0.13.1
The POM for io.sundr:sundr-codegen:jar:0.21.0 is invalid, transitive dependencies (if any) will not be available: 1 problem was encountered while building the effective model for io.sundr:sundr-codegen:0.21.0
```
The second one could be resolved by upgrading the version of the kubernetes libraries.

- cqengine relies on libraries that rely on older versions of Java (<11). This results in a warning, again resulting from javassis-3.25-GA trying to access com.sun.:tools. Unfortunately upgrading the version does not resolve this issue. See also https://stackoverflow.com/a/60549167.

**Resolved error for running the build with JDK 13 (target is still 11)** due to generic type resolution by a library used by cqengine that caused the build to fail due to failing tests.. The fix is to not use anonymous method references or lambdas in methods that call net.jodah.typetools.TypeResolver:

```java
 /**
     * The {@link com.googlecode.cqengine.query.QueryFactory#attribute(String, SimpleFunction)})} relies on a method
     * {@link net.jodah.typetools.TypeResolver#resolveRawArguments(Type, Class)}, which in Java 13 is not able to retrieve
     * information about the generic types, if a lambda or anonymous method reference is provided. By providing an anonymous
     * class of the {@link SimpleFunction}, the generic types can be resolved without running into exceptions.
     */
    @SuppressWarnings({"Convert2Lambda", "Anonymous2MethodRef"})
    private static final Attribute<Item, String> IDENTIFIER = attribute("identifier", new SimpleFunction<>() {
        @Override
        public String apply(Item item) {
            return item.getIdentifier();
        }
    });
```

**Adapted readme.md for Windows Docker support.**

**Added test steps to CI for JDK 11 and 13 on all OS**

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -58,10 +58,13 @@ Clone nivio, build and run a Docker image:
...
       mvn clean package
       docker build -t nivio:latest .
-      docker run -e SEED=/tmp/nivio/inout.yml --mount type=bind,source="C:\<your>\<path>\<to>\nivio\src\test\resources\example",target=/tmp/nivio -p 8080:8080 nivio:latest
+      docker run -e SEED=//tmp/nivio/inout.yml --mount type=bind,source="C:\<your>\<path>\<to>\nivio\src\test\resources\example",target=/tmp/nivio -p 8080:8080 nivio:latest
       
    then open http://localhost:8080
    
+   (Note: the double slashes at the beginning of the path for the SEED environment variable work as a fix to make MSYS/MinGW consoles
+   *not* translate the `/tmp` path to a local DOS path. This is safe to use with the PowerShell. Further reading: https://stackoverflow.com/a/14189687/10000398)
+
```
Comment (mfbieber) [inline readme.md]: This would cause nivio to believe the `SEED` is e.g. `C:/Users/User/AppData/Local/Temp/nivio/inout.yml`.

Comment (bonndan) [reply readme.md]: 👍

File: src/test/docker-compose.yml [hunk 2]
```diff
@@ -9,9 +9,10 @@ services:
       - 9229:9229
 
   nivio:
-    image: bonndan/nivio:latest
+    image: dedica/nivio:latest
     environment:
-      DEMO: 1
+      DEMO: ${DEMO}
+      SEED: ${SEED}
```
Comment (mfbieber) [inline src/test/docker-compose.yml:15]: By this, we could use this docker-compose.yml for both the demo and the seed config as described in the manual. @bonndan: were these docker-compose.yml files even intended for this purpose?

Comment (bonndan) [reply src/test/docker-compose.yml:15]: No, you have found an ancient relict. However, it could work that way. Using both seed and demo is now supported.

**Comment by mfbieber** [comment]
Regarding the "Seed config" in the `install.rst`, currently it says:
```rst
Seed config
-----------

Nivio expects a seed configuration at start time. You can either set an landscapeDescription variable *SEED* to a path to read from,
or you can omit the variable and place landscapeDescription files in /opt/nivio/environments.

.. code-block:: bash

    SEED=/my/files/here docker-compose up

then point your browser to http://localhost:8080/render/dld4e/{landscape} where landscape is the identifier of the landscape
to be rendered. The seed can also be an URL.
```

I was running `SEED="C:\Users\Nutzer\workspace\nivio\src\test\resources\example\example-k8s.yml" docker-compose up` forthe `docker-compose.yml` in `src/test`.  

The documentation says, I should point my browser to http://localhost:8080/render/dld4e/k8s:example, since the identifier is "k8s:example". This does not really work and I don't really understand what I am supposed to do.

Also, I cannot see any landscapes from the home path (http://localhost:8080/#/).

The console logs:

```shell
Nutzer@DESKTOP-KKSA9OM MINGW64 ~/workspace/nivio/src/test (issue-138-and-java-13-support)
$ SEED="C:\Users\Nutzer\workspace\nivio\src\test\resources\example\example-k8s.yml" docker-compose up
The DEMO variable is not set. Defaulting to a blank string.
Starting test_dld4e-headless_1 ... done
Recreating test_nivio_1        ... done
Attaching to test_dld4e-headless_1, test_nivio_1
dld4e-headless_1  | Dld4e app listening on port 3030!
nivio_1           |
nivio_1           |   .   ____          _            __ _ _
nivio_1           |  /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
nivio_1           | ( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
nivio_1           |  \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
nivio_1           |   '  |____| .__|_| |_|_| |_\__, | / / / /
nivio_1           |  =========|_|==============|___/=/_/_/_/
nivio_1           |  :: Spring Boot ::        (v2.2.7.RELEASE)
nivio_1           |
nivio_1           | 2020-09-05 19:13:55.821  INFO 1 --- [           main] de.bonndan.nivio.Application             : Starting Application v0.3.0 on 6e883bf1219d with PID 1 (/app.jar started by root in /)
nivio_1           | 2020-09-05 19:13:55.825  INFO 1 --- [           main] de.bonndan.nivio.Application             : No active profile set, falling back to default profiles: default
nivio_1           | 2020-09-05 19:13:57.164  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
nivio_1           | 2020-09-05 19:13:57.181  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
nivio_1           | 2020-09-05 19:13:57.181  INFO 1 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.34]
nivio_1           | 2020-09-05 19:13:57.264  INFO 1 --- [           main] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
nivio_1           | 2020-09-05 19:13:57.264  INFO 1 --- [           main] o.s.web.context.ContextLoader            : Root WebApplicationContext: initialization completed in 1380 ms
nivio_1           | 2020-09-05 19:13:57.470  INFO 1 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'clientInboundChannelExecutor'
nivio_1           | 2020-09-05 19:13:57.473  INFO 1 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'clientOutboundChannelExecutor'
nivio_1           | 2020-09-05 19:13:57.515  INFO 1 --- [           main] o.s.s.c.ThreadPoolTaskScheduler          : Initializing ExecutorService 'messageBrokerTaskScheduler'
nivio_1           | 2020-09-05 19:13:57.551  INFO 1 --- [           main] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService 'brokerChannelExecutor'
nivio_1           | 2020-09-05 19:13:57.885  INFO 1 --- [           main] o.s.b.a.w.s.WelcomePageHandlerMapping    : Adding welcome page: class path resource [static/index.html]
nivio_1           | 2020-09-05 19:13:58.908  INFO 1 --- [           main] o.s.m.s.b.SimpleBrokerMessageHandler     : Starting...
nivio_1           | 2020-09-05 19:13:58.908  INFO 1 --- [           main] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=true, SimpleBrokerMessageHandler [DefaultSubscriptionRegistry[cache[0 destination(s)], registry[0 sessions]]]]
nivio_1           | 2020-09-05 19:13:58.909  INFO 1 --- [           main] o.s.m.s.b.SimpleBrokerMessageHandler     : Started.
nivio_1           | 2020-09-05 19:13:58.955  INFO 1 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat started on port(s): 8080 (http) with context path ''
nivio_1           | 2020-09-05 19:13:58.962  INFO 1 --- [           main] de.bonndan.nivio.Application             : Started Application in 3.791 seconds (JVM running for 4.254)
nivio_1           | 2020-09-05 19:13:58.968  INFO 1 --- [TaskExecutor-14] de.bonndan.nivio.input.Seed              : Using seeds: [C:\Users\Nutzer\workspace\nivio\src\test\resources\example\example-k8s.yml]
nivio_1           | 2020-09-05 19:13:59.025  WARN 1 --- [TaskExecutor-16] d.b.n.notification.NotificationService   : Cannot send error, landscape is not configured
nivio_1           | 2020-09-05 19:13:59.026  INFO 1 --- [TaskExecutor-17] d.b.nivio.notification.MessagingService  : Broadcasting processing event: ProcessingErrorEvent
nivio_1           | 2020-09-05 19:13:59.889  INFO 1 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
nivio_1           | 2020-09-05 19:13:59.890  INFO 1 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
nivio_1           | 2020-09-05 19:13:59.896  INFO 1 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 6 ms
nivio_1           | 2020-09-05 19:14:03.921  INFO 1 --- [MessageBroker-1] d.b.nivio.observation.ObserverRegistry   : Polling 0 landscapes for changes.
nivio_1           | 2020-09-05 19:14:23.926  INFO 1 --- [MessageBroker-2] d.b.nivio.observation.ObserverRegistry   : Polling 0 landscapes for changes.
nivio_1           | 2020-09-05 19:14:43.926  INFO 1 --- [MessageBroker-3] d.b.nivio.observation.ObserverRegistry   : Polling 0 landscapes for changes.
nivio_1           | 2020-09-05 19:14:57.550  INFO 1 --- [MessageBroker-1] o.s.w.s.c.WebSocketMessageBrokerStats    : WebSocketSession[2 current WS(2)-HttpStream(0)-HttpPoll(0), 3 total, 0 closed abnormally (0 connect failure, 0 send limit, 0 transport error)], stompSubProtocol[processed CONNECT(3)-CONNECTED(3)-DISCONNECT(0)], stompBrokerRelay[null], inboundChannel[pool size = 21, active threads = 0, queued tasks = 0, completed tasks = 21], outboundChannel[pool size = 4, active threads = 0, queued tasks = 0, completed tasks = 4], sockJsScheduler[pool size = 5, active threads = 1, queued tasks = 1, completed tasks = 3]
nivio_1           | 2020-09-05 19:15:03.927  INFO 1 --- [MessageBroker-2] d.b.nivio.observation.ObserverRegistry   : Polling 0 landscapes for changes.

```
@bonndan: do you know what I am doing wrong? If I understand, I can also make the documentation more understandable for others to use.

File: docs/source/install.rst [hunk 3]
```diff
@@ -26,16 +26,15 @@ Set SERVER_CONTEXT_PATH and NIVIO_BASEURL env vars to the path if nivio won't be
...
 
 .. code-block:: bash
 
     DEMO=1 docker-compose up
 
 then point your browser to http://localhost:8080/render/nivio:example/graph.png
 
-There is also a demo project "nivio-demo" in the repository, which starts an nginx to simulate a remote server.
-
+There is also a demo project "nivio-demo" in the repository under `nivio/nivio-demo`, which starts an nginx to simulate a remote server.
```
Comment (mfbieber) [inline docs/source/install.rst:37]: Were those docker-compose.yml files intended for these purposes?

File: src/test/docker-compose.yml (hunk 2 above)
Comment (mfbieber) [reply src/test/docker-compose.yml:15]: Okay, so we could leave it like that? 

And on line 6 we are referencing to the image ` bonndan/dld4e-headless:latest` in this docker-compose.yml. This is your private Docker Hub user, right? Do you intend on keeping that image there?

Comment (bonndan) [reply src/test/docker-compose.yml]: OMG I was not aware the documentation is so outdated. Sorry for the confusion. I have pushed some cleanups.

File: docs/source/install.rst (hunk 3 above)
Comment (bonndan) [reply docs/source/install.rst]: I've deleted the docker-compose file under src/test and referenced the one under nivio-demo

**Comment by bonndan** [comment]
I must confess to documentation was quite outdated and I'm sorry for the confusion. The DLD4E approach is not followed anymore. I hope I have corrected everything so that the "gettings started" part is clear now. Still, there is room for improvement

File: nivio-demo/docker-compose.yml [hunk 4]
```diff
@@ -11,7 +11,9 @@ services:
 
   nivio:
     image: dedica/nivio:latest
-    #environment:
+    environment:
+      SEED: ${SEED}
+      DEMO: ${DEMO}
       # enable DEMO to have a preloaded landscape
       # DEMO: 1
       # SEED also loads data, but uses the sources from the webserver
```
Comment (mfbieber) [inline nivio-demo/docker-compose.yml:19]: To make the examples in the manual work right out of the box, we should declare the environment variables. If they are not provided, they default to an empty string. This should be no problem for our backend, right?

File: docs/source/install.rst [hunk 5]
```diff
@@ -44,16 +39,7 @@ Nivio expects a seed configuration at start time (unless you want to run the dem
...
---------------
-
-If you deploy nivio to run under a different path than root ("/"), make sure to set the environment variables
-SERVER_CONTEXT_PATH and NIVIO_BASEURL to the path.
-
-.. code-block:: bash
-
-   SERVER_SERVLET_CONTEXT_PATH: /my-landscape
-   NIVIO_BASEURL: https://foo.com/my-landscape/
+There is a demo project "nivio-demo" in the repository under `nivio/nivio-demo`, which starts an nginx to simulate a remote server.
```
Comment (mfbieber) [inline docs/source/install.rst:42]: Reading this, I still don't understand 100% how this simulates a remote server. So it simulates a remote server for the seed files? (We should write that out clearly.)

Is this the reason why my attempts to use that `docker-compose.yml` with a local file don't really work?

File: nivio-demo/docker-compose.yml (hunk 4 above)
Comment (bonndan) [reply nivio-demo/docker-compose.yml:19]: If neither SEED nor DEMO is given, nivio starts without loading any configuration. Theoretically one could push a config via rest api, but I'm not sure this is still working. Empty strings should be equal to absent vars. So yes, we should give examples.

File: docs/source/install.rst (hunk 5 above)
Comment (bonndan) [reply docs/source/install.rst:42]: I've pushed another version of the file. The demo is meant to simulate nivio running in a docker container without mounting files directly into it, and using nginx to serve configs instead.

If you have any better ideas, let me know.

File: docs/source/install.rst [hunk 6]
```diff
@@ -7,19 +7,30 @@ The Docker image is about 350MB and can be started with:
...
+
+    docker run -e DEMO=1 dedica/nivio
+
+In the demo mode Nivio loads sample data for demonstration purposes.
+
+
+There is a demo in the directory `./nivio-demo` which starts an nginx to serve all example configs from the project and
+starts nivio as docker container.
+
+From the directory, run
```
Comment (mfbieber) [inline docs/source/install.rst]: Alright, thanks. But now the `docker-compose.yml` does not have any environment variables set and when I run it, I see no landscapes in the UI. Do we still need to update the `docker-compose.yml`?

File: docs/source/install.rst [hunk 7]
```diff
@@ -58,11 +58,11 @@ bind to the corresponding folder, e.g:
        ports:
          - 8080:8080
 
-Then you can point to a specific file or folder with the *SEED* environment variable:
+Then you can point to a specific file with the *SEED* environment variable:
 
 .. code-block:: bash
 
-    SEED=/my/files/here docker-compose up
+    SEED=/my/files/here/file.yml docker-compose up
```
Comment (mfbieber) [inline docs/source/install.rst]: Providing a folder as a seed results in:
````
nivio_1      | 2020-09-13 14:24:23.492 ERROR 1 --- [TaskExecutor-14] de.bonndan.nivio.input.FileFetcher       : Failed to read file /demo
nivio_1      |
nivio_1      | java.io.IOException: Is a directory
...
````

File: nivio-demo/docker-compose.yml [hunk 8]
```diff
@@ -12,12 +12,21 @@ services:
...
+      # SEED: http://webserver:80/example_env.yml
 
+      # SEED to load data from a local file, enable also volumes
+      # SEED: /demo/dedica.yml
+    # volumes:
+      # - type: bind
+        # source: ../src/test/resources/example
+        # target: /demo
+    ports:
+      - 8080:8080
```
Comment (mfbieber) [inline nivio-demo/docker-compose.yml]: All options work on my machine now. The nginx and nivio are in the same docker network (a default one), the nginx server must be called from the inside on port 80, before we had the port on the host (3000).

Comment (bonndan) [reply nivio-demo/docker-compose.yml]: 👍

**Pull request merged**